From d23a510e1179fde91a1400fe236fa60487a21cd3 Mon Sep 17 00:00:00 2001
From: Gonzalo Odiard <godiard@gmail.com>
Date: Wed, 28 Aug 2013 12:48:47 -0300
Subject: [PATCH] Inhibit suspend while a activity is shared - OLPC #10363

This patch is a port of the patch included in the ticket.

Signed-off-by: Gonzalo Odiard <gonzalo@laptop.org>
---
 src/sugar3/activity/activity.py | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/src/sugar3/activity/activity.py b/src/sugar3/activity/activity.py
index b0339ec..2f12a0a 100644
--- a/src/sugar3/activity/activity.py
+++ b/src/sugar3/activity/activity.py
@@ -94,6 +94,8 @@ J_DBUS_SERVICE = 'org.laptop.Journal'
 J_DBUS_PATH = '/org/laptop/Journal'
 J_DBUS_INTERFACE = 'org.laptop.Journal'
 
+POWERD_INHIBIT_DIR = '/var/run/powerd-inhibit-suspend'
+
 CONN_INTERFACE_ACTIVITY_PROPERTIES = 'org.laptop.Telepathy.ActivityProperties'
 
 PREVIEW_SIZE = style.zoom(300), style.zoom(225)
@@ -774,6 +776,18 @@ class Activity(Window, Gtk.Container):
         else:
             self._jobject.metadata['share-scope'] = SCOPE_NEIGHBORHOOD
 
+    def _inhibit_suspend(self):
+        if not os.path.exists(POWERD_INHIBIT_DIR):
+            return
+
+        path = os.path.join(POWERD_INHIBIT_DIR, str(os.getpid()))
+        try:
+            fd = open(path, 'w')
+        except IOError:
+            logging.error("Inhibit Suspend: Could not create file %s", path)
+        else:
+            fd.close()
+
     def __joined_cb(self, activity, success, err):
         """Callback when join has finished"""
         logging.debug('Activity.__joined_cb %r', success)
@@ -783,6 +797,8 @@ class Activity(Window, Gtk.Container):
             logging.debug('Failed to join activity: %s', err)
             return
 
+        self._inhibit_suspend()
+
         self.reveal()
         self.emit('joined')
         self.__privacy_changed_cb(self.shared_activity, None)
@@ -811,6 +827,8 @@ class Activity(Window, Gtk.Container):
 
         activity.props.name = self._jobject.metadata['title']
 
+        self._inhibit_suspend()
+
         self.shared_activity = activity
         self.shared_activity.connect('notify::private',
                                      self.__privacy_changed_cb)
@@ -926,6 +944,11 @@ class Activity(Window, Gtk.Container):
         # Make the exported object inaccessible
         dbus.service.Object.remove_from_connection(self._bus)
 
+        if os.path.exists(POWERD_INHIBIT_DIR):
+            path = os.path.join(POWERD_INHIBIT_DIR, str(os.getpid()))
+            if os.path.exists(path):
+                os.unlink(path)
+
         self._session.unregister(self)
 
     def close(self, skip_save=False):
-- 
1.8.1.4

